File: go/internal/analysis/refactor_writer.go
// Package analysis provides static-analysis utilities including refactoring
// issue detection and report generation.
package analysis

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/unrealandychan/rekipedia/internal/models"
)

// ── Severity constants ────────────────────────────────────────────────────────

const (
	SeverityHigh   = "high"
	SeverityMedium = "medium"
	SeverityLow    = "low"
)

var severityEmoji = map[string]string{
	SeverityHigh:   "🔴",
	SeverityMedium: "🟡",
	SeverityLow:    "🟢",
}

var sectionTitle = map[string]string{
	SeverityHigh:   "High Priority",
	SeverityMedium: "Medium Priority",
	SeverityLow:    "Quick Wins (Dead Code)",
}

// godClassDegreeThreshold is the minimum combined fan-in + fan-out for a
// symbol to be flagged as a God Class.
const godClassDegreeThreshold = 10

// testPrefixes lists name prefixes that identify test helpers.
var testPrefixes = []string{"test_", "Test", "spec_", "Spec"}

// testPathSubstrings lists path fragments that identify test files.
var testPathSubstrings = []string{"/test", "\\test", "_test", "test_", "spec_", "_spec"}

// ── Data structures ───────────────────────────────────────────────────────────

// RefactorMetrics holds quantitative measurements for a detected issue.
type RefactorMetrics struct {
	Lines  int `json:"lines,omitempty"`
	FanIn  int `json:"fan_in"`
	FanOut int `json:"fan_out"`
}

// RefactorIssue represents a single detected refactoring opportunity.
type RefactorIssue struct {
	Kind       string          `json:"kind"`
	Symbol     string          `json:"symbol"`
	File       string          `json:"file"`
	Severity   string          `json:"severity"`
	Metrics    RefactorMetrics `json:"metrics"`
	Suggestion string          `json:"suggestion"`
	Callers    []string        `json:"callers"`
}

// RefactorSummary counts issues by severity.
type RefactorSummary struct {
	High   int `json:"high"`
	Medium int `json:"medium"`
	Low    int `json:"low"`
}

// RefactorReport is the machine-readable JSON output.
type RefactorReport struct {
	GeneratedAt      string          `json:"generated_at"`
	RekipediaVersion string          `json:"rekipedia_version"`
	Summary          RefactorSummary `json:"summary"`
	Issues           []RefactorIssue `json:"issues"`
}

// ── Detection logic ────────────────────────────────────────────────────────────

// DetectIssues analyses an AnalysisResult and returns detected refactoring
// issues sorted by severity (high → medium → low) and then alphabetically.
func DetectIssues(combined models.AnalysisResult) []RefactorIssue {
	var issues []RefactorIssue

	// Build degree maps
	inDeg := make(map[string]int)
	outDeg := make(map[string]int)
	callersMap := make(map[string][]string)

	for _, rel := range combined.Relationships {
		if rel.From != "" {
			outDeg[rel.From]++
		}
		if rel.To != "" {
			inDeg[rel.To]++
		}
		k := strings.ToLower(string(rel.Kind))
		if (k == "call" || k == "calls") && rel.From != "" && rel.To != "" {
			callersMap[rel.To] = append(callersMap[rel.To], rel.From)
		}
	}

	// Build symbol lookup
	symLookup := make(map[string]models.Symbol, len(combined.Symbols))
	for _, sym := range combined.Symbols {
		symLookup[sym.Name] = sym
	}

	// Entry points set
	entrySet := make(map[string]bool, len(combined.EntryPoints))
	for _, ep := range combined.EntryPoints {
		entrySet[ep] = true
	}

	// ── 1. God Class detection ─────────────────────────────────────────────
	for name, sym := range symLookup {
		kind := string(sym.Kind)
		if kind != "class" && kind != "interface" {
			continue
		}

		fanIn := inDeg[name]
		fanOut := outDeg[name]
		degree := fanIn + fanOut

		if degree < godClassDegreeThreshold {
			continue
		}

		lines := 0
		if sym.LineEnd > sym.LineStart {
			lines = sym.LineEnd - sym.LineStart
		}

		callers := uniqueStrings(callersMap[name])
		if len(callers) > 20 {
			callers = callers[:20]
		}

		issues = append(issues, RefactorIssue{
			Kind:     "god_class",
			Symbol:   name,
			File:     sym.File,
			Severity: SeverityHigh,
			Metrics: RefactorMetrics{
				Lines:  lines,
				FanIn:  fanIn,
				FanOut: fanOut,
			},
			Suggestion: fmt.Sprintf("Split `%s` into smaller, single-responsibility classes", name),
			Callers:    callers,
		})
	}

	// ── 2. Dead Code detection ─────────────────────────────────────────────
	for name, sym := range symLookup {
		kind := string(sym.Kind)
		if kind != "function" && kind != "method" {
			continue
		}
		if entrySet[name] {
			continue
		}
		// Skip test helpers
		isTestHelper := false
		for _, pfx := range testPrefixes {
			if strings.HasPrefix(name, pfx) {
				isTestHelper = true
				break
			}
		}
		if isTestHelper {
			continue
		}
		// Skip dunder methods
		if strings.HasPrefix(name, "__") && strings.HasSuffix(name, "__") {
			continue
		}
		// Skip symbols defined in test files
		isTestFile := false
		for _, sub := range testPathSubstrings {
			if strings.Contains(sym.File, sub) {
				isTestFile = true
				break
			}
		}
		if isTestFile {
			continue
		}

		if len(callersMap[name]) > 0 {
			continue
		}

		issues = append(issues, RefactorIssue{
			Kind:     "dead_code",
			Symbol:   name,
			File:     sym.File,
			Severity: SeverityLow,
			Metrics: RefactorMetrics{
				FanIn:  0,
				FanOut: outDeg[name],
			},
			Suggestion: fmt.Sprintf("Remove `%s` — 0 callers detected", name),
			Callers:    []string{},
		})
	}

	// Sort: high → medium → low, then alphabetically
	severityOrder := map[string]int{SeverityHigh: 0, SeverityMedium: 1, SeverityLow: 2}
	sort.Slice(issues, func(i, j int) bool {
		oi := severityOrder[issues[i].Severity]
		oj := severityOrder[issues[j].Severity]
		if oi != oj {
			return oi < oj
		}
		return issues[i].Symbol < issues[j].Symbol
	})

	return issues
}

// ── Markdown rendering ─────────────────────────────────────────────────────────

// BuildMarkdown renders a REFACTOR.md document from the detected issues.
func BuildMarkdown(issues []RefactorIssue, fileCount int, version string) string {
	timestamp := time.Now().UTC().Format("2006-01-02T15:04:05Z")
	n := len(issues)
	issueWord := "issues"
	if n == 1 {
		issueWord = "issue"
	}
	fileWord := "files"
	if fileCount == 1 {
		fileWord = "file"
	}

	var sb strings.Builder
	sb.WriteString("# Refactoring Guide\n")
	sb.WriteString(fmt.Sprintf("> Generated by rekipedia v%s — %s\n", version, timestamp))
	sb.WriteString(fmt.Sprintf("> %d %s found across %d %s\n\n", n, issueWord, fileCount, fileWord))

	// Group by severity preserving order
	bySeverity := map[string][]RefactorIssue{
		SeverityHigh:   {},
		SeverityMedium: {},
		SeverityLow:    {},
	}
	for _, issue := range issues {
		bySeverity[issue.Severity] = append(bySeverity[issue.Severity], issue)
	}

	for _, sev := range []string{SeverityHigh, SeverityMedium, SeverityLow} {
		sevIssues := bySeverity[sev]
		if len(sevIssues) == 0 {
			continue
		}
		emoji := severityEmoji[sev]
		title := sectionTitle[sev]
		sb.WriteString(fmt.Sprintf("## %s %s\n\n", emoji, title))

		if sev == SeverityLow {
			// Dead code — compact bulleted list
			for _, issue := range sevIssues {
				note := strings.Replace(issue.Suggestion,
					fmt.Sprintf("Remove `%s` — ", issue.Symbol), "", 1)
				prefix := ""
				if issue.File != "" {
					prefix = fmt.Sprintf("`%s:`", issue.File)
				}
				sb.WriteString(fmt.Sprintf("- %s`%s()` — %s\n", prefix, issue.Symbol, note))
			}
			sb.WriteString("\n")
		} else {
			for i, issue := range sevIssues {
				words := strings.Fields(strings.ReplaceAll(issue.Kind, "_", " "))
				for wi, w := range words {
					if len(w) > 0 {
						words[wi] = strings.ToUpper(w[:1]) + w[1:]
					}
				}
				kindLabel := strings.Join(words, " ")

				sb.WriteString(fmt.Sprintf("### %d. Split `%s` (%s)\n", i+1, issue.Symbol, kindLabel))

				var problemParts []string
				if issue.Metrics.Lines > 0 {
					problemParts = append(problemParts, fmt.Sprintf("%d lines", issue.Metrics.Lines))
				}
				if issue.Metrics.FanIn > 0 || issue.Metrics.FanOut > 0 {
					problemParts = append(problemParts,
						fmt.Sprintf("fan_in=%d, fan_out=%d", issue.Metrics.FanIn, issue.Metrics.FanOut))
				}
				problemStr := "high coupling"
				if len(problemParts) > 0 {
					problemStr = strings.Join(problemParts, ", ")
				}
				sb.WriteString(fmt.Sprintf("- **Problem**: %s\n", problemStr))
				sb.WriteString(fmt.Sprintf("- **Suggestion**: %s\n", issue.Suggestion))
				sb.WriteString(fmt.Sprintf("- **Callers affected**: %d\n", issue.Metrics.FanIn))
				if issue.File != "" {
					sb.WriteString(fmt.Sprintf("- **File**: `%s`\n", issue.File))
				}
				sb.WriteString("\n")
			}
		}
	}

	return sb.String()
}

// ── File output ────────────────────────────────────────────────────────────────

// WriteRefactorOutputs writes REFACTOR.md and refactor_report.json to outputDir.
// When stdout is true, REFACTOR.md content is also written to os.Stdout.
func WriteRefactorOutputs(combined models.AnalysisResult, outputDir string, version string, stdout bool) error {
	if err := os.MkdirAll(outputDir, 0o755); err != nil {
		return fmt.Errorf("mkdir %s: %w", outputDir, err)
	}

	issues := DetectIssues(combined)
	fileCount := len(combined.FilesSeen)

	// ── Markdown ──────────────────────────────────────────────────────────
	mdContent := BuildMarkdown(issues, fileCount, version)
	mdPath := filepath.Join(outputDir, "REFACTOR.md")
	if err := os.WriteFile(mdPath, []byte(mdContent), 0o644); err != nil {
		return fmt.Errorf("write REFACTOR.md: %w", err)
	}

	if stdout {
		if _, err := fmt.Fprint(os.Stdout, mdContent); err != nil {
			return fmt.Errorf("write REFACTOR.md to stdout: %w", err)
		}
	}

	// ── JSON ──────────────────────────────────────────────────────────────
	var summary RefactorSummary
	for _, issue := range issues {
		switch issue.Severity {
		case SeverityHigh:
			summary.High++
		case SeverityMedium:
			summary.Medium++
		case SeverityLow:
			summary.Low++
		}
	}

	// Ensure Issues is never null in JSON
	if issues == nil {
		issues = []RefactorIssue{}
	}

	report := RefactorReport{
		GeneratedAt:      time.Now().UTC().Format(time.RFC3339),
		RekipediaVersion: version,
		Summary:          summary,
		Issues:           issues,
	}

	data, err := json.MarshalIndent(report, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal refactor report: %w", err)
	}

	jsonPath := filepath.Join(outputDir, "refactor_report.json")
	if err := os.WriteFile(jsonPath, data, 0o644); err != nil {
		return fmt.Errorf("write refactor_report.json: %w", err)
	}

	return nil
}

// ── Helpers ────────────────────────────────────────────────────────────────────

// uniqueStrings returns a deduplicated copy of ss preserving first-seen order.
func uniqueStrings(ss []string) []string {
	seen := make(map[string]bool, len(ss))
	out := make([]string, 0, len(ss))
	for _, s := range ss {
		if !seen[s] {
			seen[s] = true
			out = append(out, s)
		}
	}
	return out
}

File: src/rekipedia/analysis/refactor_writer.py
"""Generate REFACTOR.md and refactor_report.json from analysis data.

This module detects refactoring opportunities from an ``AnalysisResult`` and
produces two output artefacts:

* ``REFACTOR.md``  — human/agent-readable Markdown guide
* ``refactor_report.json`` — machine-readable JSON report
"""
from __future__ import annotations

import json
import sys
from collections import defaultdict
from datetime import UTC, datetime
from pathlib import Path

from rekipedia.models.contracts import AnalysisResult, RefactorIssue, Symbol

# ── Severity → presentation mapping ─────────────────────────────────────────

_SEVERITY_EMOJI: dict[str, str] = {
    "high": "🔴",
    "medium": "🟡",
    "low": "🟢",
}

_SECTION_TITLE: dict[str, str] = {
    "high": "High Priority",
    "medium": "Medium Priority",
    "low": "Quick Wins (Dead Code)",
}

# ── Detection thresholds ─────────────────────────────────────────────────────

# A class is considered a "god class" when its combined degree reaches this value.
_GOD_CLASS_DEGREE_THRESHOLD = 10

# Test-related name prefixes/substrings — used to exclude test helpers from
# dead-code detection.
_TEST_PREFIXES = ("test_", "Test", "spec_", "Spec")
_TEST_PATH_SUBSTRINGS = ("/test", "\\test", "_test", "test_", "spec_", "_spec")


def _rekipedia_version() -> str:
    """Return the installed rekipedia package version string."""
    try:
        from importlib.metadata import version as _pkg_version

        return _pkg_version("rekipedia")
    except Exception:
        return "unknown"


def detect_issues(combined: AnalysisResult) -> list[RefactorIssue]:
    """Detect refactoring issues from an ``AnalysisResult``.

    Detects:
    * **god_class** (high severity) — classes/interfaces whose combined
      fan-in + fan-out degree exceeds ``_GOD_CLASS_DEGREE_THRESHOLD``.
    * **dead_code** (low severity) — functions/methods with zero callers that
      are not entry-points, test helpers, or dunder methods.

    Args:
        combined: ``AnalysisResult`` produced by the scan pipeline.

    Returns:
        Sorted list of issue dicts, each containing:
        ``kind``, ``symbol``, ``file``, ``severity``, ``metrics``,
        ``suggestion``, and ``callers``.
    """
    issues: list[RefactorIssue] = []

    # ── Build fan-in / fan-out maps ───────────────────────────────────────────
    fan_in: dict[str, int] = defaultdict(int)   # incoming call edges per symbol
    fan_out: dict[str, int] = defaultdict(int)  # outgoing call edges per symbol
    callers_map: dict[str, list[str]] = defaultdict(list)

    for rel in combined.relationships:
        if rel.from_:
            fan_out[rel.from_] += 1
        if rel.to:
            fan_in[rel.to] += 1
        if str(rel.kind) in ("call", "calls") and rel.from_ and rel.to:
            callers_map[rel.to].append(rel.from_)

    # ── Build symbol lookup ───────────────────────────────────────────────────
    sym_lookup: dict[str, Symbol] = {sym.name: sym for sym in combined.symbols}
    entry_points: set[str] = set(combined.entry_points)

    # ── 1. God Class detection ────────────────────────────────────────────────
    for name, sym in sym_lookup.items():
        kind_str = str(sym.kind)
        if kind_str not in ("class", "interface"):
            continue

        sym_fan_in = fan_in[name]
        sym_fan_out = fan_out[name]
        degree = sym_fan_in + sym_fan_out

        if degree < _GOD_CLASS_DEGREE_THRESHOLD:
            continue

        line_start = sym.line_start or 0
        line_end = sym.line_end or 0
        lines = max(0, line_end - line_start)

        unique_callers = list(dict.fromkeys(callers_map.get(name, [])))
        issues.append(RefactorIssue(
            kind="god_class",
            symbol=name,
            file=sym.file,
            severity="high",
            metrics={"lines": lines, "fan_in": sym_fan_in, "fan_out": sym_fan_out},
            suggestion=f"Split `{name}` into smaller, single-responsibility classes",
            callers=unique_callers[:20],
        ))

    # ── 2. Dead Code detection ────────────────────────────────────────────────
    for name, sym in sym_lookup.items():
        kind_str = str(sym.kind)
        if kind_str not in ("function", "method"):
            continue
        if name in entry_points:
            continue
        # Exclude test helpers and dunder methods
        if any(name.startswith(p) for p in _TEST_PREFIXES):
            continue
        if name.startswith("__") and name.endswith("__"):
            continue
        # Also skip symbols defined in test files
        if any(s in sym.file for s in _TEST_PATH_SUBSTRINGS):
            continue

        if callers_map.get(name):
            continue

        issues.append(RefactorIssue(
            kind="dead_code",
            symbol=name,
            file=sym.file,
            severity="low",
            metrics={"fan_in": 0, "fan_out": fan_out[name]},
            suggestion=f"Remove `{name}` — 0 callers detected",
            callers=[],
        ))

    # ── Sort: high → medium → low, then alphabetically ───────────────────────
    _order = {"high": 0, "medium": 1, "low": 2}
    issues.sort(key=lambda x: (_order.get(x.severity, 99), x.symbol))

    return issues


def _build_markdown(issues: list[RefactorIssue], file_count: int) -> str:
    """Render REFACTOR.md content from the detected issues list.

    Args:
        issues: Sorted list of issue dicts from ``detect_issues``.
        file_count: Number of files seen in the scan (used in the header).

    Returns:
        Full Markdown string ready to be written to disk.
    """
    ver = _rekipedia_version()
    timestamp = datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ")
    n_issues = len(issues)

    lines: list[str] = [
        "# Refactoring Guide",
        f"> Generated by rekipedia v{ver} — {timestamp}",
        f"> {n_issues} issue{'s' if n_issues != 1 else ''} found across "
        f"{file_count} file{'s' if file_count != 1 else ''}",
        "",
    ]

    by_severity: dict[str, list[RefactorIssue]] = {"high": [], "medium": [], "low": []}
    for issue in issues:
        sev = issue.severity or "low"
        by_severity.setdefault(sev, []).append(issue)

    for sev in ("high", "medium", "low"):
        sev_issues = by_severity.get(sev, [])
        if not sev_issues:
            continue

        emoji = _SEVERITY_EMOJI[sev]
        title = _SECTION_TITLE[sev]
        lines.append(f"## {emoji} {title}")
        lines.append("")

        if sev == "low":
            # Dead code — compact bulleted list
            for issue in sev_issues:
                sym = issue.symbol
                file_ = issue.file or ""
                note = issue.suggestion.replace(f"Remove `{sym}` — ", "")
                prefix = f"`{file_}:`" if file_ else ""
                lines.append(f"- {prefix}`{sym}()` — {note}")
            lines.append("")
        else:
            for i, issue in enumerate(sev_issues, 1):
                sym = issue.symbol
                kind = issue.kind or ""
                file_ = issue.file or ""
                metrics = issue.metrics or {}
                suggestion = issue.suggestion or ""
                fan_in = metrics.get("fan_in", 0)
                fan_out = metrics.get("fan_out", 0)
                lines_count = metrics.get("lines", 0)

                kind_label = kind.replace("_", " ").title()
                lines.append(f"### {i}. Split `{sym}` ({kind_label})")

                problem_parts: list[str] = []
                if lines_count:
                    problem_parts.append(f"{lines_count} lines")
                if fan_in or fan_out:
                    problem_parts.append(f"fan_in={fan_in}, fan_out={fan_out}")
                problem_str = ", ".join(problem_parts) if problem_parts else "high coupling"

                lines.append(f"- **Problem**: {problem_str}")
                lines.append(f"- **Suggestion**: {suggestion}")
                lines.append(f"- **Callers affected**: {fan_in}")
                if file_:
                    lines.append(f"- **File**: `{file_}`")
                lines.append("")

    return "\n".join(lines) + "\n"


def write_refactor_outputs(
    combined: AnalysisResult,
    output_dir: Path,
    *,
    stdout: bool = False,
) -> tuple[Path, Path]:
    """Write ``REFACTOR.md`` and ``refactor_report.json`` to *output_dir*.

    Args:
        combined: ``AnalysisResult`` from the scan pipeline.
        output_dir: Directory to write output artefacts into.
        stdout: When ``True``, also print ``REFACTOR.md`` content to
            ``sys.stdout`` (useful for piping to Claude Code or other tools).

    Returns:
        Tuple of ``(refactor_md_path, refactor_json_path)``.
    """
    output_dir.mkdir(parents=True, exist_ok=True)

    file_count = len(combined.files_seen)
    issues = detect_issues(combined)

    # ── Markdown ──────────────────────────────────────────────────────────────
    md_content = _build_markdown(issues, file_count)
    md_path = output_dir / "REFACTOR.md"
    md_path.write_text(md_content, encoding="utf-8")

    if stdout:
        sys.stdout.write(md_content)
        sys.stdout.flush()

    # ── JSON ──────────────────────────────────────────────────────────────────
    ver = _rekipedia_version()
    summary: dict[str, int] = {"high": 0, "medium": 0, "low": 0}
    for issue in issues:
        sev = issue.severity or "low"
        summary[sev] = summary.get(sev, 0) + 1

    report: dict = {
        "generated_at": datetime.now(UTC).isoformat(),
        "rekipedia_version": ver,
        "summary": summary,
        "issues": [i.to_dict() for i in issues],
    }
    json_path = output_dir / "refactor_report.json"
    json_path.write_text(
        json.dumps(report, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )

    return md_path, json_path
